pkg/models/sigv4.go

```diff
@@ -0,0 +1,78 @@
+package models
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+)
+
+type AuthType string
+
+const (
+	Keys        AuthType = "keys"
+	Credentials AuthType = "credentials"
+	ARN         AuthType = "arn"
+)
+
+type Sigv4Middleware struct {
+	Config *Config
+	Next   http.RoundTripper
+}
+
+type Config struct {
+	AuthType string
+
+	Profile string
+
+	AssumeRoleARN string
+	ExternalID    string
+
+	AccessKey string
+	SecretKey string
+
+	Region string
+}
+
+func (m *Sigv4Middleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	_, err := m.signRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if m.Next == nil {
+		return http.DefaultTransport.RoundTrip(req)
+	}
+
+	return m.Next.RoundTrip(req)
+}
+
+func (m *Sigv4Middleware) signRequest(req *http.Request) (http.Header, error) {
+	creds := m.credentials()
+	if creds == nil {
+		return nil, errors.New("invalid credentials option")
+	}
+
+	signer := v4.NewSigner(creds)
+	return signer.Sign(req, nil, "grafana", m.Config.Region, time.Now())
+}
+
+func (m *Sigv4Middleware) credentials() *credentials.Credentials {
+	authType := AuthType(m.Config.AuthType)
+
+	switch authType {
+	case Keys:
+		return credentials.NewStaticCredentials(m.Config.AccessKey, m.Config.SecretKey, "")
+	case Credentials:
+		return credentials.NewSharedCredentials("", m.Config.Profile)
+	case ARN:
+		s := session.Must(session.NewSession())
```

Comment:
Not entirely sure if this is the most correct way to achieve what we need here. The session has a timeout which is defaulted to 15m - maybe we can be more aggressive. I'm waiting also for the rework of the current Cloudwatch auth work to be completed - see [here](https://github.com/grafana/grafana/pull/25548)

Reply:
Following my above comment check how the role assumption is suggested to be implemented [here](https://github.com/grafana/grafana/blob/34fb224ec7e794d4b4b284ead20b857d5bc7b501/pkg/tsdb/cloudwatch/cloudwatch.go#L126)

Reply:
Yeah, I do think the session `s` should be from one of the authentication methods. I think how it's created here, it will use the default SDK method implicitly?

Reply:
Yep that's what I gathered from my testing of it too @aknuds1